# DTSC 580: Data Manipulation
## Assignment: LEGO Project
### Name: 

![LegoSchema](Lego_Schema.PNG)

## Overview

In this assignment, you will load the various tables from the [Rebrickable](https://rebrickable.com/) website and perform analysis on the data.  Rebrickable is a fantastic website that is used to show you which LEGO sets you can build from the sets and parts you already own.  They will even show you which parts you are missing to complete the new set and suggest sets that you could purchase to get those parts.  

## Data

The data has been downloaded directly from their website and includes files for the tables shown in the schema above.  Your first task will be to go through the data and review the schema to understand how the tables all fit together.  There is no data dictionary that I am aware of, but you should be able to understand this data by reviewing the files and schema and visiting their [Download](https://rebrickable.com/downloads/) page and [API](https://rebrickable.com/api/) documentation.  **Note that you must use the files downloaded from Brightspace for this assignment and not download new files from the Rebrickable website.**

## Assignment

In this assignment, you will create a number of functions to complete certain tasks.  This is not needed if you were working on this project on your own, but since we want to see if you can select the data via Pandas/Numpy, we will use those functions in the automatic grading in CodeGrade.  Without this, students could try to look up the answers in the `csv` files instead of wrangling the data with Python/Pandas, which would defeat the purpose of this class.

In addition, we would strongly suggest that you merge data sets as appropriate for the given task.  For example, you might merge the `sets` and `themes` DataFrames to answer one question, and then merge `sets`, `inventories`, `inventory_minifigs`, and `minifigs` to answer another.  This will make your code much easier to write and will be key in working with data across multiple tables.  In fact, I was able to create each function with only one line of code because of the merging of certain data sets.  Also note that you may not use all the tables in this assignment.  This is also an important skill for a Data Scientist to know when to use data and what data is unnecessary.

When complete, save your notebook as `lego.ipynb` and submit to CodeGrade for automatic grading.

## Note

<u>Show Work</u>

Remember that you must show your work.  Students submissions are spot checked manually throughout the term to verify that they are not hard coding the answer from looking only in the file or in CodeGrade's expected output.  If this is seen, the student's answer will be manually marked wrong and their grade will be changed to reflect this. 

For example, if the answer to Q1, the mean of a specific column, is 22:
```
# correct way
Q1 = df['column_name'].mean()

# incorrect way
Q1 = 22 
```

Let's start out by importing some standard imports.

In [1]:
# common imports
import pandas as pd
import numpy as np

# Do not change this option; This allows the CodeGrade auto grading to function correctly
pd.set_option('display.max_columns', None)

## Import Data

Load the various Lego datasets below calling them the same as the name of the table as seen in the schema.  Note that you may or may not have to use all the data sets.  One of the things that all Data Scientists and Analysts have to decide when working on a project is what data is useful for a task and what data is not.  Remember to look at the schema above to see how all the data sets are connected.

The autograder in CodeGrade will have the `csv` files in the same location where the notebook is run.  Because of this, please make sure your local files are in the same location as your notebook.

In [2]:
# import Lego data sets here
# make sure that files are in the same folder as your notebook
inventories = pd.read_csv('inventories.csv')
inventory_sets = pd.read_csv('inventory_sets.csv')
sets = pd.read_csv('sets.csv')
themes = pd.read_csv('themes.csv')
inventory_minifigs = pd.read_csv('inventory_minifigs.csv')
minifigs = pd.read_csv('minifigs.csv')
inventory_parts = pd.read_csv('inventory_parts.csv')
part_categories = pd.read_csv('part_categories.csv')
parts = pd.read_csv('parts.csv')
colors = pd.read_csv('colors.csv')
part_relationships = pd.read_csv('part_relationships.csv')
elements = pd.read_csv('elements.csv')

## Exploration

Take time to look at all of the data sets using `.head()`, `.info()`, `.describe()`, etc to familiarize yourself with the data sets.  As you look at them, notice how the various tables connect together.

In [3]:
### ENTER CODE HERE ###
inventories 


id  version        set_num
0           1        1         7922-1
1           3        1         3931-1
2           4        1         6942-1
3          15        1         5158-1
4          16        1          903-1
...       ...      ...            ...
31582  122328        2          426-1
31583  122329        1  210-2-DUMMY-1
31584  122344        2         7036-1
31585  122450        1        11961-1
31586  122472        3        10262-1

[31587 rows x 3 columns]

In [4]:
inventory_sets 


inventory_id  set_num  quantity
0               35  75911-1         1
1               35  75912-1         1
2               39  75048-1         1
3               39  75053-1         1
4               50   4515-1         1
...            ...      ...       ...
3408        118961  71032-7         1
3409        118961  71032-8         1
3410        118961  71032-9         1
3411        122079  71032-0         6
3412        122081  71032-0        36

[3413 rows x 3 columns]

In [5]:
sets 


set_num                             name  year  theme_id  num_parts
0            001-1                            Gears  1965         1         43
1           0011-2                Town Mini-Figures  1979        67         12
2           0011-3       Castle 2 for 1 Bonus Offer  1987       199          0
3           0012-1               Space Mini-Figures  1979       143         12
4           0013-1               Space Mini-Figures  1979       143         12
...            ...                              ...   ...       ...        ...
18571      XWING-1              Mini X-Wing Fighter  2019       158         60
18572      XWING-2                X-Wing Trench Run  2019       158         52
18573  YODACHRON-1  Yoda Chronicles Promotional Set  2013       158        413
18574   YTERRIER-1                Yorkshire Terrier  2018       598          0
18575     ZX8000-1             ZX 8000 LEGO Sneaker  2020       501          0

[18576 rows x 5 columns]

In [6]:
themes 


id                        name  parent_id
0      1                     Technic        NaN
1      3                 Competition        1.0
2      4              Expert Builder        1.0
3     16                  RoboRiders        1.0
4     17              Speed Slammers        1.0
..   ...                         ...        ...
439  715                      Marvel      535.0
440  716                     Modulex        NaN
441  717                 Speed Racer        NaN
442  718       Series 22 Minifigures      535.0
443  719  BrickLink Designer Program        NaN

[444 rows x 3 columns]

In [7]:
inventory_minifigs 


inventory_id     fig_num  quantity
0                 3  fig-001549         1
1                 4  fig-000764         1
2                19  fig-000555         1
3                25  fig-000574         1
4                26  fig-000842         1
...             ...         ...       ...
18139        122344  fig-009900         1
18140        122344  fig-009908         2
18141        122344  fig-009954         1
18142        122344  fig-009955         1
18143        122344  fig-009956         1

[18144 rows x 3 columns]

In [8]:
minifigs 


fig_num                                   name  num_parts
0      fig-000001                     Toy Store Employee          4
1      fig-000002                           Customer Kid          4
2      fig-000003                  Assassin Droid, White          8
3      fig-000004                           Basic Figure          4
4      fig-000005        Captain America with Short Legs          3
...           ...                                    ...        ...
11822  fig-012221                                Bob-omb         11
11823  fig-012222                   Boomer (Banzai) Bill         10
11824  fig-012223                            Bullet Bill         10
11825  fig-012224                           Buzzy Beetle          7
11826  fig-012225  Unikitty - Cheerykitty (Cheery Kitty)         14

[11827 rows x 3 columns]

In [9]:
inventory_parts 


inventory_id        part_num  color_id  quantity is_spare
0                  1        48379c01        72         1        f
1                  1           48395         7         1        f
2                  1  stickerupn0077      9999         1        f
3                  1         upn0342         0         1        f
4                  1         upn0350        25         1        f
...              ...             ...       ...       ...      ...
989524        122472           99206        71         2        f
989525        122472           99207        71         2        f
989526        122472           99207         0         2        f
989527        122472           99780         0         2        f
989528        122472           99780         1         1        f

[989529 rows x 5 columns]

In [10]:
part_categories


id                     name
0    1               Baseplates
1    3            Bricks Sloped
2    4  Duplo, Quatro and Primo
3    5           Bricks Special
4    6            Bricks Wedged
..  ..                      ...
61  64      Minidoll Lower Body
62  65         Minifig Headwear
63  66                  Modulex
64  67   Tiles Round and Curved
65  68  Projectiles / Launchers

[66 rows x 2 columns]

In [11]:
parts 


part_num                                        name  part_cat_id  \
0           003381                 Sticker Sheet for Set 663-1           58   
1           003383         Sticker Sheet for Sets 618-1, 628-2           58   
2           003402  Sticker Sheet for Sets 310-3, 311-1, 312-3           58   
3           003429                Sticker Sheet for Set 1550-1           58   
4           003432  Sticker Sheet for Sets 357-1, 355-1, 940-1           58   
...            ...                                         ...          ...   
44776  znapupn0002       Znap Connector 3 x 3 - 4 way B (Beam)           43   
44777  znapupn0003              Znap Connector 1 x 3 - 2 way A           43   
44778  znapupn0004                         Znap Beam 3, 1 Hole           43   
44779  znapupn0005     Znap Connector 3 x 3 - 4 way C (Closed)           43   
44780  znapupn0006           Wheel 68mm Znap Propeller (9 x 2)           29   

      part_material  
0           Plastic  
1           Plastic  
2           Plastic  
3           Plastic  
4           Plastic  
...             ...  
44776       Plastic  
44777       Plastic  
44778       Plastic  
44779       Plastic  
44780       Plastic  

[44781 rows x 4 columns]

In [12]:
colors 


id                  name     rgb is_trans
0      -1             [Unknown]  0033B2        f
1       0                 Black  05131D        f
2       1                  Blue  0055BF        f
3       2                 Green  237841        f
4       3        Dark Turquoise  008F9B        f
..    ...                   ...     ...      ...
185  1059     Trans-Purple Opal  8320B7        t
186  1060      Trans-Green Opal  84B68D        t
187  1061  Trans-Dark Blue Opal  0020A0        t
188  1062        Vibrant Yellow  EBD800        f
189  9999  [No Color/Any Color]  05131D        f

[190 rows x 4 columns]

In [13]:
part_relationships 


rel_type child_part_num parent_part_num
0            P    87079pr9974           87079
1            P     3960pr9971            3960
2            R    98653pr0003     98086pr0003
3            R    98653pr0003    98088pat0003
4            R    98653pr0003    98089pat0003
...        ...            ...             ...
24503        P    98138pr0284           98138
24504        B    98138pr0284           83079
24505        P    11203pr9998           11203
24506        P    98138pr0285           98138
24507        B    98138pr0285           83079

[24508 rows x 3 columns]

In [14]:
elements

element_id     part_num  color_id
0        6300211     67906c01        14
1        4566309         2564         0
2        4275423        53657      1004
3        6194308        92926        71
4        6229123        26561         4
...          ...          ...       ...
62679    6288007        66429      9999
62680    4622289  92755pr0001        72
62681    6375358  98138pr0283       297
62682    3894199         3894        72
62683    6380659        88293       297

[62684 rows x 3 columns]

## Assignment

Answer all the questions below using the available data.  There are multiple ways to get some of the answers.  I would suggest that you attempt to merge some data sets as needed to assist with getting answers.  This will make your tasks a lot easier and your code will be cleaner.

Also, you will be creating functions to get the answers to some of these questions.  This will make it much easier for you to get answers to similar questions in the future and CodeGrade will be checking your functions using different inputs.  

**Q1:** Create a function called `data_count` that takes as input the data set (for example: `sets`) and returns the total number of rows in the data (as an integer).  Nothing else should be returned.

In [15]:
### ENTER CODE HERE ###
def data_count(df):
  return len(df.index)

**Code Check:**  If you call your `data_count` function using the `sets` DataFrame, you should see that there are 18,576 sets represented in the data.

In [16]:
### CODE CHECK ###
data_count(sets)


18576

**Q2:** What is the average number of parts for a Lego set (rounded to the nearest integer)?  Save this as `Q2`.

In [17]:
### ENTER CODE HERE ###
Q2 = round(sets['num_parts'].mean())
Q2

161

**Q3:** What is the median number of parts per set (output as an integer)?  Save this as `Q3`.

In [18]:
### ENTER CODE HERE ###
Q3 = round(sets['num_parts'].median())
Q3

37

You should see that the median number of parts is much lower than the average number of parts per set.  This tells us that there are some Lego sets with a very high number of parts that brings the average up.

**Q4:** What is the largest number of parts that a set has? Save this as `Q4`.

In [19]:
### ENTER CODE HERE ###
Q4 = round(sets['num_parts'].max())
Q4

11695

**Q5:** Given the set with the largest number of parts calculated above, select its row from the `sets` DataFrame.  Save this as `Q5`.  Do not reset the index.

In [20]:
### ENTER CODE HERE ###
Q5 = sets[sets['num_parts']==Q4]
Q5

set_num       name  year  theme_id  num_parts
3867  31203-1  World Map  2021       709      11695

**Q6:** Create a function called `select_set_row` where you pass it the `set_num` (as a string) and the function returns the respective row of the DataFrame from `sets`. Do not reset the index.

In [21]:
### ENTER CODE HERE ###
def select_set_row(set_num):
  return sets[sets['set_num']==set_num]

select_set_row('0012-1')

set_num                name  year  theme_id  num_parts
3  0012-1  Space Mini-Figures  1979       143         12

**Q7:** Create a function called `select_set_numparts` that takes as input a specific number of parts (as an integer) and selects all the rows in the `sets` DataFrame that contain those number of parts.  Return the DataFrame sorted by `year` and then by `set_num`.  Do not reset the index.

In [22]:
### ENTER CODE HERE ###
def select_set_numparts(num):
  return sets[sets['num_parts']==num].sort_values(by=['year','set_num'])



**Code Check:** Call the `select_set_numparts()` function to select all sets with `0` number of parts.  There should be 3,146 rows returned.

In [23]:
### CODE CHECK ###
select_set_numparts(0)

set_num                                               name  \
11061          700.1-2                     Extra-Large Gift Set (Mursten)   
17563         b55dk-01               Lego Mursten - System i Leg Byggebog   
17564         b56de-01                                    System im Spiel   
11072          700.2-2                                       Gift Package   
11078         700.3A-2                                       Gift Package   
...                ...                                                ...   
15619         854195-1                                    Ernie Key Chain   
15620         854197-1                               Demogorgon Key Chain   
17170  9780744054583-1  DC Comics Super Heroes: Character Encyclopedia...   
17193  9780794447557-1                               City: Color The City   
17200  9780794448172-1           DC Comics Super Heroes: Ready For Action   

       year  theme_id  num_parts  
11061  1953       366          0  
17563  1955       497          0  
17564  1956       497          0  
11072  1957       366          0  
11078  1957       372          0  
...     ...       ...        ...  
15619  2022       503          0  
15620  2022       503          0  
17170  2022       497          0  
17193  2022       497          0  
17200  2022       497          0  

[3146 rows x 5 columns]

**Q8:** 
- The oldest set listed was made in what year?  Save the year as `Q8A` (as an integer).  
- The newest set listed was made in what year? Save the year as `Q8B` (as an integer).

In [24]:
### ENTER CODE HERE ###
Q8A = sets['year'].min()
Q8A

1949

In [25]:
### ENTER CODE HERE ###
Q8B = sets['year'].max()
Q8B

2022

**Q9:** 
- Create a function called `select_set_year` that takes as input a year (as an integer) and returns the rows from the `sets` DataFrame that matches that year.  Sort the DataFrame by `set_num`.  Do not reset the index.
- Use the above function to return a DataFrame for rows that match the minimum year (calculated as `Q8A` above) and call this `Q9A`.  
- Then use the above function to return a DataFrame for rows that match the maximum year (calculated as `Q8B` above) and call this `Q9B`.

In [26]:
### ENTER CODE HERE ###
def select_set_year(year):
  return sets[sets['year']==year].sort_values(by='set_num')

In [27]:
### ENTER CODE HERE ###
Q9A = select_set_year(Q8A)
Q9A 

set_num                               name  year  theme_id  num_parts
11057  700.1-1         Extra-Large Gift Set (ABB)  1949       365        142
11070  700.2-1               Large Gift Set (ABB)  1949       365        178
11075  700.3-1              Medium Gift Set (ABB)  1949       365        142
11080  700.A-1              Small Brick Set (ABB)  1949       371         24
11081  700.B-1  Small Doors and Windows Set (ABB)  1949       371         12

In [28]:
### ENTER CODE HERE ###
Q9B = select_set_year(Q8B)
Q9B 

set_num                                               name  \
336            10297-1                                     Boutique Hotel   
600            10781-1                          Spider-Man’s Techno Trike   
601            10782-1                      Hulk vs. Rhino Truck Showdown   
602            10783-1                        Spider-Man at Doc Ock’s Lab   
603            10784-1                     Spider-Man Webquarters Hangout   
...                ...                                                ...   
17170  9780744054583-1  DC Comics Super Heroes: Character Encyclopedia...   
17193  9780794447557-1                               City: Color The City   
17200  9780794448172-1           DC Comics Super Heroes: Ready For Action   
17207  9780794449247-1  DC Comics Super Heroes: Gotham City’s New Defe...   
17726    DOTS-SERIES-6                    DOTS Collectible Tiles Series 6   

       year  theme_id  num_parts  
336    2022       155       3068  
600    2022       706         59  
601    2022       706        110  
602    2022       706        131  
603    2022       706        155  
...     ...       ...        ...  
17170  2022       497          0  
17193  2022       497          0  
17200  2022       497          0  
17207  2022       497          7  
17726  2022       688         16  

[183 rows x 5 columns]

**Q10:** What is the median year represented in the `sets` data?  Save this as `Q10` (as an integer).

In [29]:
### ENTER CODE HERE ###
Q10 = round(sets['year'].median())
Q10

2010

Notice that the first set was back in 1949 but over half of the sets were after the median year. This could be an indication that the number of sets has increased greatly over the past number of years.

**Q11:** What is the average number of sets listed per year?  Save this as `Q11` (as an integer).

In [30]:
### ENTER CODE HERE ###
Q11 = round(sets.groupby('year')['set_num'].count().mean())
Q11

258

**Q12:** What year had the lowest total number of sets?  Save this year as `Q12` (as an integer).

In [31]:
### ENTER CODE HERE ###
Q12 = round(sets.groupby('year').size().sort_values().index[0])
Q12

1960

**Q13:** What year had the highest total number of sets? Save this as `Q13` (as an integer).

In [32]:
### ENTER CODE HERE ###
Q13 = round(sets.groupby('year').size().sort_values(ascending=False).index[0])
Q13

2021

**Q14:** 
- Create a function called `theme_by_year` that takes as input a year (as an integer) and shows the theme ids and theme names (listed in order by theme id) that were in sets that year.  
- The column names must be `id` and `name_themes` (to differentiate between the name of a theme and the name of a set) in that order.  
- The index should be reset and go from `0` to `n-1`.  
- Each theme should only be listed once even if it appeared in more than one set from that year -- duplicate themes should be based on theme id and not name since there are some themes with the same name but with a different id.
- Hint:  It will help if you were to think about merging appropriate DataFrames to help you get this answer.

In [33]:
### ENTER CODE HERE ###
sets_with_theme = sets.merge(themes, left_on='theme_id', right_on='id', suffixes=('_sets','_themes'))
def theme_by_year(year):
  year_df = sets_with_theme[sets_with_theme['year']==year]
  year_df = year_df[['id','name_themes']].drop_duplicates().sort_values(by='id').reset_index(drop=True)
  return year_df


**Code Check:** Call the `theme_by_year()` function on the year with the lowest number of total sets as found in `Q12`.  Your output should look as follows:

| | id | name_themes|
|-|---|---|
|0|371|Supplemental|
|1|497|Books|
|2|513|Classic|

**Optional Practice:** Call the `theme_by_year()` function using your birth year to see the Lego themes that were popular when you were born.

In [34]:
### CODE CHECK / OPTIONAL PRACTICE ###
theme_by_year(Q12)

id   name_themes
0  371  Supplemental
1  497         Books
2  513       Classic

**Q15:** What theme has the most number of sets associated with it?  Save the theme name as `Q15` (as a string).

In [35]:
### ENTER CODE HERE ###
Q15 = sets_with_theme.groupby(['theme_id','name_themes'])['set_num'].count().sort_values(ascending=False).index[0][1]
Q15

'Gear'

**Q16:** 
- Create a function called `theme_by_name` that takes as input a theme name and outputs all the sets (in `set_num` order) that had that theme.  
- The DataFrame should include the following columns in this order and with these exact names: `set_num`,`name_sets`,`year`,`num_parts`, `theme_id`, and `name_themes`.  
- The index should go from `0` to `n-1`.

In [36]:
### ENTER CODE HERE ###
def theme_by_name(theme_name):
  df = sets_with_theme[sets_with_theme['name_themes']==theme_name].sort_values(by='set_num')
  df = df[['set_num','name_sets','year','num_parts','theme_id','name_themes']]

  return df.reset_index(drop=True)

**Code Check:** Call the `theme_by_name()` function using the theme name from `Q15`.  You should return 1,904 sets that match that theme name.

In [37]:
### CODE CHECK ###
theme_by_name(Q15)

set_num                                          name_sets  \
0              11138-1                            Bionicle Barraki Wallet   
1       119645111436-1                Classic Space Astronaut Plush (Red)   
2       119645111504-1               Classic Space Astronaut Plush (Blue)   
3       119645111740-1              Classic Space Astronaut Plush (White)   
4       119645111818-1               Classic Space Astronaut Plush (Pink)   
...                ...                                                ...   
1899           WATCH-3                   Watch System Voyager (UFO Space)   
1900         WIIGAME-1                              Bionicle Heroes - Wii   
1901         WIIGAME-2  Indiana Jones: The Original Adventures Video G...   
1902  WOODENELEPHANT-1  Original Wooden Elephant Pull-Along Toy and Print   
1903          ZX8000-1                               ZX 8000 LEGO Sneaker   

      year  num_parts  theme_id name_themes  
0     2008          0       501        Gear  
1     2021          0       501        Gear  
2     2021          0       501        Gear  
3     2021          0       501        Gear  
4     2021          0       501        Gear  
...    ...        ...       ...         ...  
1899  1997         29       501        Gear  
1900  2007          0       501        Gear  
1901  2008          0       501        Gear  
1902  2021          0       501        Gear  
1903  2020          0       501        Gear  

[1904 rows x 6 columns]

**Q17:** What themes have the most sets associated with them?  Output the top 5 themes as a Series with the theme name as the index and the number of sets that the theme appears in as the values.  The Series should be sorted in descending order.  Save this as `Q17`.

In [38]:
### ENTER CODE HERE ###
Q17 = sets_with_theme.groupby('name_themes').size().nlargest(5)
Q17

name_themes
Gear         1904
Star Wars     825
Technic       693
Key Chain     583
Friends       482
dtype: int64

**Q18:** Create a function called `theme_by_setnum` that takes as input the set_num (as a string) and returns the respective theme name as a string.

In [39]:
### ENTER CODE HERE ###
def theme_by_setnum(set_num):
  theme = sets_with_theme[sets_with_theme['set_num']==set_num]['name_themes']
  return theme.values[0]


**Code Check:** Visit the [Sets page](https://rebrickable.com/sets/) from the Rebrickable website and find a set that you like.  Call your `theme_by_setnum()` function on that set and make sure the correct theme name is output.

In [40]:
### CODE CHECK ###
theme_by_setnum('WATCH-3')

'Gear'

**Q19:** Find all the sets named 'Millennium Falcon'. Of these sets, find the set with the largest number of parts.  You **do not** need to save this as `Q19`.  Instead you will use this information for later questions.

In [41]:
### ENTER CODE HERE ###
sets_with_theme[sets_with_theme['name_sets']=='Millennium Falcon'].nlargest(1,columns='num_parts')
#largest set 75192-1
Q19='75192-1'

**Q20:** 
- How many minifigs are contained in the set from `Q19`?  Save this as `Q20` (as an integer).  
- Note that there are sometimes multiple versions of a set as shown in the `inventories` table.  This will not matter for this question as there is only one version for the respective set but this should be noted for future questions.
- Remember to show your work or your answer will manually get marked as incorrect.

In [42]:
### ENTER CODE HERE ###
big_merged_table = sets_with_theme.merge(inventories, on='set_num', suffixes=('_themes','_inventory'))
big_merged_table = big_merged_table.merge(inventory_minifigs, left_on='id_inventory', right_on='inventory_id')
big_merged_table = big_merged_table.merge(minifigs, on='fig_num', suffixes=('_sets','_fig'))
#big_merged_table

In [43]:
Q20 = big_merged_table[big_merged_table['set_num']==Q19].groupby(['version','fig_num']).size().sum()
Q20

8

In [44]:
big_merged_table[big_merged_table['set_num']==Q19]


set_num          name_sets  year  theme_id  num_parts_sets  id_themes  \
1791  75192-1  Millennium Falcon  2017       171            7541        171   
2564  75192-1  Millennium Falcon  2017       171            7541        171   
2584  75192-1  Millennium Falcon  2017       171            7541        171   
2635  75192-1  Millennium Falcon  2017       171            7541        171   
2640  75192-1  Millennium Falcon  2017       171            7541        171   
2650  75192-1  Millennium Falcon  2017       171            7541        171   
2651  75192-1  Millennium Falcon  2017       171            7541        171   
2652  75192-1  Millennium Falcon  2017       171            7541        171   

                    name_themes  parent_id  id_inventory  version  \
1791  Ultimate Collector Series      158.0         19670        1   
2564  Ultimate Collector Series      158.0         19670        1   
2584  Ultimate Collector Series      158.0         19670        1   
2635  Ultimate Collector Series      158.0         19670        1   
2640  Ultimate Collector Series      158.0         19670        1   
2650  Ultimate Collector Series      158.0         19670        1   
2651  Ultimate Collector Series      158.0         19670        1   
2652  Ultimate Collector Series      158.0         19670        1   

      inventory_id     fig_num  quantity  \
1791         19670  fig-002057         1   
2564         19670  fig-001714         1   
2584         19670  fig-002514         1   
2635         19670  fig-001810         1   
2640         19670  fig-002544         1   
2650         19670  fig-002549         1   
2651         19670  fig-002550         1   
2652         19670  fig-002551         1   

                                                   name  num_parts_fig  
1791                                 Rey, Dark Tan Robe              4  
2564                              Chewbacca, Dark Brown              3  
2584    C-3PO, Pearl Gold, Colorful Wires, Printed Legs              3  
2635                             Finn, Black Undershirt              4  
2640                                               BB-8              2  
2650  Princess Leia, Hoth Outfit, White Jacket, Prin...              4  
2651  Han Solo, Dark Blue Jacket, Dark Brown Legs, P...              4  
2652                               Han Solo, Old, Angry              4

**Q21:** 
- Create a function called `get_minifigs` that takes as input a set number (as a string) and a version number (as an integer) and outputs all minifigs from that set / version combination.  
- The DataFrame should be in `fig_num` order with the following columns/column names: `set_num`, `fig_num`, `fig_name`, `fig_num_parts`.  
- The index should go from 0 to n-1.

In [45]:
### ENTER CODE HERE ###
def get_minifigs(set_num, version=1):
  mini = big_merged_table[big_merged_table['set_num']==set_num]
  mini = mini[mini['version']==version]
  mini = mini.rename({'name':'fig_name','num_parts_fig':'fig_num_parts'},axis=1).sort_values(by='fig_num').reset_index()[['set_num','fig_num','fig_name','fig_num_parts']]
  return mini

**Q22:** Using the above function, output information about the minifigs from the set discussed in `Q19`.  Save this output as `Q22`.

In [46]:
### ENTER CODE HERE ###
Q22 = get_minifigs(Q19)
Q22

set_num     fig_num                                           fig_name  \
0  75192-1  fig-001714                              Chewbacca, Dark Brown   
1  75192-1  fig-001810                             Finn, Black Undershirt   
2  75192-1  fig-002057                                 Rey, Dark Tan Robe   
3  75192-1  fig-002514    C-3PO, Pearl Gold, Colorful Wires, Printed Legs   
4  75192-1  fig-002544                                               BB-8   
5  75192-1  fig-002549  Princess Leia, Hoth Outfit, White Jacket, Prin...   
6  75192-1  fig-002550  Han Solo, Dark Blue Jacket, Dark Brown Legs, P...   
7  75192-1  fig-002551                               Han Solo, Old, Angry   

   fig_num_parts  
0              3  
1              4  
2              4  
3              3  
4              2  
5              4  
6              4  
7              4

**Code Check:** Make sure that your `get_minifigs()` function works with both version 1 and version 2 of set 10217-1, Harry Potter's Diagon Alley.  Both versions should have 11 minifigs in the DataFrame.

In [47]:
### CODE CHECK ###
get_minifigs('10217-1',2)

set_num     fig_num                                           fig_name  \
0   10217-1  fig-002229                   Skeleton (Standard) (3626b Head)   
1   10217-1  fig-006075                 Rubeus Hagrid, Dark Brown Overcoat   
2   10217-1  fig-006084  Lucius Malfoy, Black Robes and Cape, Death Eat...   
3   10217-1  fig-006130  Harry Potter, Open Dark Blue Jacket over White...   
4   10217-1  fig-006138                                    Fenrir Greyback   
5   10217-1  fig-006162  Ron Weasley, Red Tartan Sweater Vest, Reddish ...   
6   10217-1  fig-006163          Fred / George Weasley in Dark Orange Suit   
7   10217-1  fig-006164                                 Garrick Ollivander   
8   10217-1  fig-006165    Goblin, Black Jacket and Light Bluish Gray Legs   
9   10217-1  fig-006166        Goblin, Black Jacket and Reddish Brown Legs   
10  10217-1  fig-006167  Hermione Granger, Light Bluish Gray Sweater wi...   

    fig_num_parts  
0               6  
1               3  
2               5  
3               4  
4               4  
5               4  
6               4  
7               4  
8               3  
9               3  
10              4

**Q23:** 
- Create a function called `minifigs_from_themes` that takes as input a theme id (as an integer) and outputs all the minifigs contained in sets with that theme.  
- The DataFrame should be in `fig_num` order with the following columns/column names: `fig_num`, `fig_name`. 
- The index should go from 0 to n-1.  
- Exact duplicates should be removed so that only unique figs are included.

In [48]:
### ENTER CODE HERE ###
def minifigs_from_themes(id):
  df = big_merged_table[big_merged_table['theme_id']==id].rename({'name':'fig_name'}, axis=1).sort_values(by='fig_num')
  

  return df[['fig_num','fig_name']].drop_duplicates().reset_index(drop=True)


**Code Check:** Call the `minifigs_from_themes()` function using the theme number of the set discussed in `Q19`.  There should be 149 minifigs output for that respective theme.

In [49]:
### CODE CHECK ###
minifigs_from_themes(171)

fig_num                                           fig_name
0    fig-000003                              Assassin Droid, White
1    fig-000124                           Chewbacca, Reddish Brown
2    fig-000144  Luke Skywalker, White Robe, White Legs, Cheek ...
3    fig-000303           Clone Trooper, Phase I Armor, Brown Eyes
4    fig-000516  Darth Vader, Light Bluish Gray Skin, Scarred R...
..          ...                                                ...
144  fig-011942  AT-AT Driver - Light Bluish Grey Helmet, Sand ...
145  fig-011944  Snowtrooper, White Hips, Reddish Brown Head, F...
146  fig-011945  Snowtrooper, White Hips, Light Flesh Head, Female
147  fig-011946  Snowtrooper, White Hips, Light Flesh Head, Off...
148  fig-011957    Snowtrooper, White Hips, Light Flesh Head, Grin

[149 rows x 2 columns]

**Q24:** 
- Create a function called `sets_from_minifig` that takes as input a minifig number (as a string) and returns a list of sets a minifig has appeared in.  
- Return a DataFrame (sorted by `set_num` and then by `version`) with the following columns: `set_num`, `version`,`name_sets`, `fig_num`, `fig_name`.  
- The index should go from 0 to n-1.

In [50]:
### ENTER CODE HERE ###
def sets_from_minifig(fig_num):
  df = big_merged_table[big_merged_table['fig_num']==fig_num].rename({'name':'fig_name'},axis=1)

  return df.sort_values(by=['set_num','version']).reset_index()[['set_num','version','name_sets','fig_num','fig_name']]

**Code Check:** Call your `sets_from_minifig()` function on one of the Lego Friends minifigs, Andrea, who has figure number: `fig-001039`.  Your function should return a DataFrame that contains the following sets:

- Andrea on the Beach
- Birthday Party
- City Park Cafe

In [51]:
### CODE CHECK ###
sets_from_minifig('fig-001039')

set_num  version            name_sets     fig_num  \
0  30100-1        1  Andrea on the Beach  fig-001039   
1  30107-1        1       Birthday Party  fig-001039   
2   3061-1        1       City Park Café  fig-001039   

                                            fig_name  
0  Andrea - Bright Light Orange Top, Light Aqua S...  
1  Andrea - Bright Light Orange Top, Light Aqua S...  
2  Andrea - Bright Light Orange Top, Light Aqua S...

**Q25:** 
- Create a function called `set_parts` that takes as input the set number (as a string) and version number (as an integer) and outputs a DataFrame with each part in the set.  
- The DataFrame should have the following columns/column names: `set_num`, `version`, `name_sets`, `year`, `name_themes`, `part_num`, `name_part`, `name_color`,`quantity`,`is_spare`.  
- The Data should be sorted by `part_num` and then `name_color`.  
- The index should go from 0 to n-1.

In [52]:
big_merged_table = big_merged_table.merge(inventory_parts, on='inventory_id', suffixes=('_fig','_part')).rename({'name':'fig_name'},axis=1)
big_merged_table = big_merged_table.merge(colors, left_on='color_id', right_on='id').drop(columns=['id','id_inventory','id_themes']).rename({'name':'name_color'},axis=1)
big_merged_table = big_merged_table.merge(parts, on='part_num').rename({'name':'name_part'},axis=1)

In [53]:
### ENTER CODE HERE ###
def set_parts(set_num, version=1):
  df = big_merged_table[big_merged_table['set_num']==set_num]
  df = df[df['version']==version].rename({'quantity_part':'quantity'},axis=1)
  df = df[['set_num','version','name_sets','year','name_themes','part_num','name_part','name_color','quantity','is_spare']].drop_duplicates().reset_index(drop=True)

  return df.sort_values(['part_num','name_color'])

**Code Check**: Using the `set_parts()` function, output a list of parts for the set discussed in `Q19`. There should be a total of 730 rows in the DataFrame, which includes 676 parts and 54 spare parts. You can check the respective page on the rebrickable website to verify these results.

In [54]:
### CODE CHECK ###
set_parts(Q19)

set_num  version          name_sets  year                name_themes  \
644  75192-1        1  Millennium Falcon  2017  Ultimate Collector Series   
560  75192-1        1  Millennium Falcon  2017  Ultimate Collector Series   
538  75192-1        1  Millennium Falcon  2017  Ultimate Collector Series   
537  75192-1        1  Millennium Falcon  2017  Ultimate Collector Series   
634  75192-1        1  Millennium Falcon  2017  Ultimate Collector Series   
..       ...      ...                ...   ...                        ...   
601  75192-1        1  Millennium Falcon  2017  Ultimate Collector Series   
396  75192-1        1  Millennium Falcon  2017  Ultimate Collector Series   
397  75192-1        1  Millennium Falcon  2017  Ultimate Collector Series   
518  75192-1        1  Millennium Falcon  2017  Ultimate Collector Series   
699  75192-1        1  Millennium Falcon  2017  Ultimate Collector Series   

    part_num                                          name_part  \
644    10197  Technic Pin Connector Hub with 2 Perpendicular...   
560    10201            Bracket 1 x 2 - 1 x 4 [Rounded Corners]   
538    10247  Plate Special 2 x 2 with 1 Pin Hole [Complete ...   
537    10247  Plate Special 2 x 2 with 1 Pin Hole [Complete ...   
634    10314  Brick Curved 1 x 4 x 1 1/3 No Studs, Curved To...   
..       ...                                                ...   
601    99563  Tile Special 1 x 2 with Sloped Walls AKA Money...   
396    99780                     Bracket 1 x 2 - 1 x 2 Inverted   
397    99780                     Bracket 1 x 2 - 1 x 2 Inverted   
518    99781                              Bracket 1 x 2 - 1 x 2   
699    99784  Bar 1 x 12 with 1 x 2 Plate End with Hollow St...   

            name_color  quantity is_spare  
644   Dark Bluish Gray         3        f  
560  Light Bluish Gray        12        f  
538   Dark Bluish Gray         5        f  
537              White        22        f  
634                Tan         3        f  
..                 ...       ...      ...  
601  Light Bluish Gray        70        f  
396              Black         6        f  
397   Dark Bluish Gray        25        f  
518  Light Bluish Gray         2        f  
699  Light Bluish Gray         3        f  

[730 rows x 10 columns]

  **Q26:**  Get a list of all unique color names, sort by color name, and output as a list.  Save this as `Q26`. 

In [55]:
### ENTER CODE HERE ###
Q26 = big_merged_table['name_color'].drop_duplicates().values.tolist()
Q26

['Black',
 'Yellow',
 'Dark Gray',
 'Light Gray',
 'Red',
 'White',
 'Blue',
 'Reddish Brown',
 'Dark Turquoise',
 'Dark Bluish Gray',
 'Dark Orange',
 'Medium Azure',
 'Dark Blue',
 'Trans-Red',
 'Trans-Dark Blue',
 'Green',
 'Brown',
 'Light Bluish Gray',
 'Trans-Neon Orange',
 'Sand Green',
 'Orange',
 'Dark Tan',
 'Dark Green',
 'Dark Red',
 'Lime',
 'Dark Purple',
 'Magenta',
 'Tan',
 'Dark Azure',
 'Metallic Silver',
 'Trans-Clear',
 'Medium Blue',
 'Bright Green',
 'Dark Pink',
 'Pink',
 'Dark Brown',
 'Pearl Gold',
 'Medium Dark Flesh',
 'Bright Pink',
 'Sand Blue',
 'Bright Light Orange',
 'Coral',
 'Light Aqua',
 'Bright Light Blue',
 'Medium Lavender',
 'Lavender',
 'Bright Light Yellow',
 'Vibrant Yellow',
 'Blue-Violet',
 'Earth Orange',
 'Milky White',
 'Maersk Blue',
 'Yellowish Green',
 'Pastel Blue',
 'Medium Orange',
 'Flat Silver',
 'Purple',
 'Glow in Dark White',
 'Pearl Light Gray',
 'Olive Green',
 'Trans-Yellow',
 'Trans-Green',
 'Trans-Light Blue',
 'Trans-Neon

**Q27:** How many total transparent colors are in the `colors` DataFrame?  Save this as `Q27` (as an integer).

In [56]:
### ENTER CODE HERE ###
Q27 = colors[colors['is_trans']=='t']['id'].count()

#Q27=0
#for i in Q26:
#  if 'trans' in i.lower():
#    Q27+=1
#Q27

**Q28:**  
- Create a function called `get_part_colors` that takes as input a part number (as a string) and returns a DataFrame of all colors a part has appeared in.  
- The output will return the corresponding rows from the `colors` DataFrame for each color.
- The DataFrame should be sorted by color id  
- **Do not reset the index.**

In [57]:
### ENTER CODE HERE ###
def get_part_colors(part_num):
  df = colors.merge(inventory_parts, left_on='id', right_on='color_id').drop(columns='color_id').merge(parts, on='part_num', suffixes=(None,'_part'))
  df = df[df['part_num']==part_num]


  return df[['id','name','rgb','is_trans']].sort_values(['id'])

In [58]:
get_part_colors('3437')

id                      name     rgb is_trans
2       -1                 [Unknown]  0033B2        f
70       0                     Black  05131D        f
69       0                     Black  05131D        f
68       0                     Black  05131D        f
67       0                     Black  05131D        f
...    ...                       ...     ...      ...
3842  1050                     Coral  FF698F        f
3843  1050                     Coral  FF698F        f
3844  1050                     Coral  FF698F        f
3845  1050                     Coral  FF698F        f
3847  1057  Trans-Light Bright Green  C9E788        t

[3846 rows x 4 columns]

## Optional Exercises
These exercises are optional and are not graded by CodeGrade.  They can be used as extra optional practice.

- What minifig has the most number of parts?
- What minifig shows up in the most number of sets?
- When was the first minifig included in a set?
- What set has the most number of minifigs
- What part is used the most for all sets?
- What color is used the most for all parts?
- What part / color combination is used the most for all sets?
- What part / color combination is used the least for all sets?
- Explore some of your favorite themes and which colors are used the most for those themes.
- Plot a line graph of the average number of parts per set per year.
